### PR TITLE
Updates to Artena local plugin

### PR DESCRIPTION
**Pull request by daemonhunt**
Changes made to bring the Artena-Moodle plugin up-to-date with the 2.8+ and 2.9+ versions of Moodle, mostly fixes to deprecated code. May want to create a separate branch for these. May also want to update the versioning of each of the files to keep this version distinct. I welcome any tests/comments.

### CODE REVIEW

File: local/artena/externallib.php [hunk 1]
```diff
@@ -182,9 +197,9 @@ public static function create_course($courses) {
         require_once($CFG->libdir . '/completionlib.php');
 
         $params = self::validate_parameters(self::create_course_parameters(),
-                        array('courses' => $courses));
+            array('courses' => $courses));
 
-        $availablethemes = get_plugin_list('theme');
+        $availablethemes = core_component::get_plugin_list('theme');
```
Comment (daemonhunt) [inline local/artena/externallib.php:202]: Deprecated code fix

File: local/artena/externallib.php [hunk 2]
```diff
@@ -192,7 +207,7 @@ public static function create_course($courses) {
         foreach ($params['courses'] as $course) {
 
             // Ensure the current user is allowed to run this function
-            $context = get_context_instance(CONTEXT_COURSECAT, $course['categoryid']);
+            $context = context_coursecat::instance($course['categoryid']);
```
Comment (daemonhunt) [inline local/artena/externallib.php:210]: Deprecated code fix

File: local/artena/externallib.php [hunk 3]
```diff
@@ -214,17 +229,17 @@ public static function create_course($courses) {
                 require_capability('moodle/course:create', $context);
 
                 // Make sure lang is valid
-                if (key_exists('lang', $course) and empty($availablelangs[$course['lang']])) {
+                if (array_key_exists('lang', $course) and empty($availablelangs[$course['lang']])) {
```
Comment (daemonhunt) [inline local/artena/externallib.php:232]: Deprecated code fix

File: local/artena/externallib.php [hunk 4]
```diff
@@ -214,17 +229,17 @@ public static function create_course($courses) {
...
                     throw new moodle_exception(
-                            get_string('errorinvalidparam', 'webservice', 'lang'));
+                        get_string('errorinvalidparam', 'webservice', 'lang'));
                 }
 
                 // Make sure theme is valid
-                if (key_exists('forcetheme', $course)) {
+                if (array_key_exists('forcetheme', $course)) {
```
Comment (daemonhunt) [inline local/artena/externallib.php:238]: Deprecated code fix

File: local/artena/externallib.php [hunk 5]
```diff
@@ -240,10 +255,10 @@ public static function create_course($courses) {
                 //set default value for completion
                 $courseconfig = get_config('moodlecourse');
                 if (completion_info::is_enabled_for_site()) {
-                    if (!key_exists('enablecompletion', $course)) {
+                    if (!array_key_exists('enablecompletion', $course)) {
```
Comment (daemonhunt) [inline local/artena/externallib.php:258]: Deprecated code fix

File: local/artena/externallib.php [hunk 6]
```diff
@@ -240,10 +255,10 @@ public static function create_course($courses) {
...
                         $course['enablecompletion'] = $courseconfig->enablecompletion;
                     }
-                    if (!key_exists('completionstartonenrol', $course)) {
+                    if (!array_key_exists('completionstartonenrol', $course)) {
```
Comment (daemonhunt) [inline local/artena/externallib.php:261]: Deprecated code fix

File: local/artena/externallib.php [hunk 7]
```diff
@@ -355,7 +375,7 @@ public static function change_course_id($courses) {
             } else {
 
                 // Ensure the current user is allowed to run this function
-                $context = get_context_instance(CONTEXT_COURSECAT, $existing_course->category);
+                $context = context_coursecat::instance($existing_course->category);
```
Comment (daemonhunt) [inline local/artena/externallib.php:378]: Deprecated code fix

File: local/artena/externallib.php [hunk 8]
```diff
@@ -447,11 +472,11 @@ public static function remove_course($courses) {
                 $existing_course = $DB->get_record('course', array('fullname' => $course['fullname'], 'shortname' => $course['shortname']));
             }
             if (false === $existing_course) {   // unknown course
-                throw new invalid_parameter_exception('Unknown course: '.$assignment['idnumber']);
+                throw new invalid_parameter_exception('Unknown course: '.$course['idnumber']);
```
Comment (daemonhunt) [inline local/artena/externallib.php:475]: Unsure if this is correct... $assignment should be $course? Haven't had any trouble with it so far.

File: local/artena/externallib.php [hunk 9]
```diff
@@ -447,11 +472,11 @@ public static function remove_course($courses) {
...
             } else {
 
                 // Ensure the current user is allowed to run this function
-                $context = get_context_instance(CONTEXT_COURSECAT, $existing_course->category);
+                $context = context_coursecat::instance($existing_course->category);
```
Comment (daemonhunt) [inline local/artena/externallib.php:479]: Deprecated code fix

File: local/artena/externallib.php [hunk 10]
```diff
@@ -739,7 +772,7 @@ public static function create_user_parameters() {
                             'lang'        => new external_value(PARAM_SAFEDIR, 'Language code such as "en", must exist on server', VALUE_DEFAULT, $CFG->lang, NULL_NOT_ALLOWED),
                             'theme'       => new external_value(PARAM_SAFEDIR, 'Theme name such as "standard", must exist on server', VALUE_OPTIONAL),
                             'timezone'    => new external_value(PARAM_ALPHANUMEXT, 'Timezone code such as Australia/Perth, or 99 for default', VALUE_OPTIONAL),
-                            'mailformat'  => new external_value(PARAM_INTEGER, 'Mail format code is 0 for plain text, 1 for HTML etc', VALUE_OPTIONAL),
+                            'mailformat'  => new external_value(PARAM_INT, 'Mail format code is 0 for plain text, 1 for HTML etc', VALUE_OPTIONAL),
```
Comment (daemonhunt) [inline local/artena/externallib.php:775]: PARAM_INT

File: local/artena/externallib.php [hunk 11]
```diff
@@ -756,7 +789,7 @@ public static function create_user_parameters() {
                                         'type'  => new external_value(PARAM_ALPHANUMEXT, 'The name of the custom field'),
                                         'value' => new external_value(PARAM_RAW, 'The value of the custom field')
                                     )
-                                ), 'User custom fields (also known as user profil fields)', VALUE_OPTIONAL)
+                                ), 'User custom fields (also known as user profile fields)', VALUE_OPTIONAL)
```
Comment (daemonhunt) [inline local/artena/externallib.php:792]: Minor spelling error

File: local/artena/externallib.php [hunk 12]
```diff
@@ -767,29 +800,34 @@ public static function create_user_parameters() {
...
         require_once($CFG->dirroot."/user/lib.php");
         require_once($CFG->dirroot."/user/profile/lib.php"); //required for customfields related function
-                                                             //TODO: move the functions somewhere else as
-                                                             //they are "user" related
+        //TODO: move the functions somewhere
+        //they are "user" related
 
         // Ensure the current user is allowed to run this function
-        $context = get_context_instance(CONTEXT_SYSTEM);
+        $context = context_system::instance();
```
Comment (daemonhunt) [inline local/artena/externallib.php:819]: Deprecated code fix

File: local/artena/externallib.php [hunk 13]
```diff
@@ -767,29 +800,34 @@ public static function create_user_parameters() {
...
         self::validate_context($context);
         require_capability('moodle/user:create', $context);
 
         // Do basic automatic PARAM checks on incoming data, using params description
         // If any problems are found then exceptions are thrown with helpful error messages
         $params = self::validate_parameters(self::create_user_parameters(), array('users'=>$users));
-        $availableauths  = get_plugin_list('auth');
+        $availableauths  = core_component::get_plugin_list('auth');
```
Comment (daemonhunt) [inline local/artena/externallib.php:826]: Deprecated code fix

File: local/artena/externallib.php [hunk 14]
```diff
@@ -767,29 +800,34 @@ public static function create_user_parameters() {
...
         unset($availableauths['mnet']);       // these would need mnethostid too
         unset($availableauths['webservice']); // we do not want new webservice users for now
 
-        $availablethemes = get_plugin_list('theme');
+        $availablethemes = core_component::get_plugin_list('theme');
```
Comment (daemonhunt) [inline local/artena/externallib.php:830]: Deprecated code fix

File: local/artena/externallib.php [hunk 15]
```diff
@@ -887,46 +925,48 @@ public static function create_user($users) {
         return $userids;
     }
 
-	public static function create_user_record($user) {
-		global $DB;
+    public static function create_user_record($user, $triggerevent = true) {
```
Comment (daemonhunt) [inline local/artena/externallib.php:928]: Deprecated code fix for triggerevent

File: local/artena/externallib.php [hunk 16]
```diff
@@ -887,46 +925,48 @@ public static function create_user($users) {
...
-		// insert the user into the database
-		$newuserid = $DB->insert_record('user', $user);
+        // insert the user into the database
+        $newuserid = $DB->insert_record('user', $user);
 
-		// trigger user_created event on the full database user row
-		$newuser = $DB->get_record('user', array('id' => $newuserid));
-		events_trigger('user_created', $newuser);
+        // trigger user_created event on the full database user row if required
+        $newuser = $DB->get_record('user', array('id' => $newuserid));
```
Comment (daemonhunt) [inline local/artena/externallib.php]: Triggerevent deprecated code fix

File: local/artena/externallib.php [hunk 17]
```diff
@@ -887,46 +925,48 @@ public static function create_user($users) {
...
+        if ($triggerevent) {
```
Comment (daemonhunt) [inline local/artena/externallib.php:955]: Triggerevent deprecated code fix

File: local/artena/externallib.php [hunk 18]
```diff
@@ -887,46 +925,48 @@ public static function create_user($users) {
...
+            \core\event\user_created::create_from_userid($newuserid)->trigger();
+        }
 
-		// create USER context for this user
-		get_context_instance(CONTEXT_USER, $newuserid);
+        // create USER context for this user
+        context_user::instance($newuserid);
```
Comment (daemonhunt) [inline local/artena/externallib.php:960]: Deprecated code fix

File: local/artena/externallib.php [hunk 19]
```diff
@@ -967,22 +1007,26 @@ public static function remove_user_parameters() {
...
+     * @throws required_capability_exception
+     * @throws restricted_context_exception
      */
     public static function remove_user($users) {
         global $CFG, $DB;
         require_once($CFG->dirroot."/user/lib.php");
 
         // Ensure the current user is allowed to run this function
-        $context = get_context_instance(CONTEXT_SYSTEM);
+        $context = context_system::instance();
```
Comment (daemonhunt) [inline local/artena/externallib.php:1022]: Deprecated code fix

File: local/artena/externallib.php [hunk 20]
```diff
@@ -967,22 +1007,26 @@ public static function remove_user_parameters() {
...
         self::validate_context($context);
         require_capability('moodle/user:delete', $context);
 
         // Do basic automatic PARAM checks on incoming data, using params description
         // If any problems are found then exceptions are thrown with helpful error messages
         $params = self::validate_parameters(self::remove_user_parameters(), array('users'=>$users));
-
+        $availableauths  = core_component::get_plugin_list('auth');
```
Comment (daemonhunt) [inline local/artena/externallib.php:1029]: availableauths string did not appear to be defined

File: local/artena/externallib.php [hunk 21]
```diff
@@ -1090,21 +1140,21 @@ public static function create_enrol($assignments) {
...
+                    throw new invalid_parameter_exception('Unknown course: '.$assignment['idnumber']);
                 }
-              } else {
-                throw new invalid_parameter_exception('Unknown course: '.$assignment['idnumber']);
-              }
             }
 
 //self::log_for_artena('create_enrol','get context');
-            $context = get_context_instance(CONTEXT_COURSE, $existing_course->id);
+            $context = context_course::instance($existing_course->id);
```
Comment (daemonhunt) [inline local/artena/externallib.php:1157]: Deprecated code fix

File: local/artena/externallib.php [hunk 22]
```diff
@@ -1280,7 +1337,7 @@ public static function remove_enrol($assignments) {
                 throw new invalid_parameter_exception('Unknown course: '.$assignment['idnumber']);
             }
 
-            $context = get_context_instance(CONTEXT_COURSE, $existing_course->id);
+            $context = context_course::instance($existing_course->id);
```
Comment (daemonhunt) [inline local/artena/externallib.php:1340]: Deprecated code fix

File: local/artena/externallib.php [hunk 23]
```diff
@@ -1333,9 +1390,109 @@ public static function remove_enrol_returns() {
         return null;
     }
 
+
```
Comment (daemonhunt) [inline local/artena/externallib.php:1393]: This section was included in plugin obtained from SMSS, however, wasn't included in GitHub

File: local/artena/externallib.php [hunk 24]
```diff
@@ -1333,9 +1390,109 @@ public static function remove_enrol_returns() {
...
+                )
+            )
+        );
+    }
+
     public static function log_for_artena($method,$data,$new=0) {
-	// comment this out to enable logging
-	return;
+        // comment this out to enable logging
+
```
Comment (daemonhunt) [inline local/artena/externallib.php:1495]: return; removed

File: local/artena/lang/en/local_artena.php [hunk 25]
```diff
@@ -0,0 +1,30 @@
+<?php
```
Comment (daemonhunt) [inline local/artena/lang/en/local_artena.php:1]: This language file ensures that a correct menu entry is created for the local plugin. Doesn't break plugin without it, just makes it a bit more user friendly and tidy. All files need reversioning.

**Comment by danmarsden** [comment]
it looks like the majority of these fixes have been implemented already - it would be good if you could verify this and close this pull request if completed.

**Pull request closed without merging**